opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -0,0 +1,641 @@
+            }
+        }
+        return result;
+    }
+}
+
+void extendImportList (std::vector<std::tuple<int,int,char,int>>& importList,
+                                const std::vector<std::set<int>>& extraWells)
+{
+    using ImportList = std::vector<std::tuple<int,int,char,int>>;
+    // make a list of wells for easy identification. Contains ID, begin, end
+    using iter = std::set<int>::const_iterator;
+    std::unordered_map<int,std::pair<iter,iter>> wellMap;
+    for (const auto& well : extraWells)
+    {
+        if (well.size()>1)
+        {
+            wellMap[*well.begin()] = std::make_pair(well.begin(),well.end());
+        }
+    }
+
+    ImportList addToList;
+    // iterate once through the original importList
+    for (const auto& cellProperties : importList)
+    {
+        // if a cell is in any well, add cells of the well to importList
+        auto pWell = wellMap.find(std::get<0>(cellProperties));
+        if (pWell!=wellMap.end())
+        {
+            const auto& [begin,end] = pWell->second;
+            // well ID is its cell of lowest index and is already in the importList
+            assert(*begin==std::get<0>(cellProperties));
+            for (auto pgID = begin; ++pgID!=end; )
+            {
+                // cells in one well have the same attributes (except ID)
+                std::tuple<int,int,char,int> wellCell = cellProperties;
+                std::get<0>(wellCell) = *pgID;
+                addToList.push_back(wellCell);
+            }
+
+            wellMap.erase(pWell);
+            if (wellMap.empty())
+            {
+                break;
+            }
+        }
+    }
+
+    // add new cells to the importList and sort it. It is assumed that importList starts sorted.
+    auto compareTuple = [](const auto& a, const auto& b){return std::get<0>(a)<std::get<0>(b);};
+    std::sort(addToList.begin(),addToList.end(),compareTuple);
+    auto origSize = importList.size();
+    auto totsize = origSize+addToList.size();
+    importList.reserve(totsize);
+    importList.insert(importList.end(),addToList.begin(),addToList.end());
+    std::inplace_merge(importList.begin(),importList.begin()+origSize,importList.end(),compareTuple);
+}
+
+} // end namespace Impl
+
+void extendExportAndImportLists(const GraphOfGrid<Dune::CpGrid>& gog,
+         const Dune::cpgrid::CpGridDataTraits::Communication& cc,
+                                                          int root,
+                       std::vector<std::tuple<int,int,char>>& exportList,
+                   std::vector<std::tuple<int,int,char,int>>& importList,
+                                      const std::vector<int>& gIDtoRank)
+{
+    // extend root's export list and get sets of well cells for other ranks
+    auto expListToComm = Impl::extendRootExportList(gog,exportList,root,gIDtoRank);
+    // obtain wells on this rank from root
+    auto extraWells = Impl::communicateExportedWells(expListToComm,cc,root);
```

Comment:
There is a bit of a mismatch between variable name and function name. Seeing both at once I can get what the function does. Can we think of a better name? Maybe getWellCellFromRoot?

Not explicit here, but this is a `vector<set<int>>`. Do we need the entries to be in a set below? Otherwise it might be better to use vector<vector>> (Processing received data in communicateExportedWells might become easier and faster).

Reply:
The function extendRootExportList iterates over the export list and adds well cells to the export list. The wells in the export list are those that need to be communicated to non-root ranks, and we return them as vectors of sets.

The function does two things so I did not come with a better name, now it explains just one half of its purpose. extendRootExportListAndReturnNonrootWells seemed a little too long.

Using vectors instead of sets is indeed an option, I will add it to my todo list.

Reply:
Fair enough.